### PR TITLE
KAZOO-5039 modb pagination

### PR DESCRIPTION
**Pull request by jamhed**

### CODE REVIEW

File: applications/crossbar/src/crossbar_doc.erl [hunk 1]
```diff
@@ -475,6 +477,28 @@ load_view(#load_view_params{view=View
...
+map_options(CurrentKey, Options) ->
+    OptStartKey = props:get_value('startkey', Options),
+    OptEndKey = props:get_value('endkey', Options),
+    Descending = props:get_value('descending', Options),
+    map_options(CurrentKey, Descending, OptStartKey, OptEndKey, Options).
+
+-spec map_options(term(), term(), term(), term(), kz_proplist()) -> kz_proplist().
+map_options(CurrentKey, 'true', _, _, Options) ->
+    CleanOpts = props:delete_keys(['startkey', 'descending'], Options),
+    props:fiter_undefined([{'startkey', CurrentKey}, 'descending' | CleanOpts]);
```
Comment (fenollp) [inline applications/crossbar/src/crossbar_doc.erl]: typo

File: applications/crossbar/src/crossbar_doc.erl [hunk 2]
```diff
@@ -381,22 +381,28 @@ load_view(View, Options, Context, StartKey, PageSize) ->
...
 view_sort_direction(Options) ->
-    case props:get_value('descending', Options) of
-        'true' -> 'descending';
-        'undefined' -> 'ascending'
+    Descending = props:get_value('descending', Options),
+    StartKey = props:get_value('startkey', Options),
+    EndKey = props:get_value('endkey', Options),
+    case {Descending, StartKey, EndKey} of
+        {'true', _, _} -> 'descending';
+        {_, StartKey, EndKey} when is_integer(StartKey), is_integer(EndKey), StartKey < EndKey -> 'descending';
```
Comment (fenollp) [inline applications/crossbar/src/crossbar_doc.erl]: Are these ever integers?

Comment (jamhed) [inline applications/crossbar/src/crossbar_doc.erl]: @fenollp, sometimes they are.

File: applications/crossbar/src/cb_pager.erl [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+limited_query(Limit, _, _, _) when Limit < 0; Limit == 0 -> [];
+limited_query(Limit, Db, View, CouchOpts) ->
+    case kz_datamgr:get_results(Db, View, [{'limit', Limit} | CouchOpts]) of
+        {'ok', JObjs} -> JObjs;
+        {'error', 'not_found'} -> [];
+        {'error', Error} -> throw(Error)
+    end.
+
+-spec apply_filter({arity, integer()}, fun(), list()) -> list().
+apply_filter({'arity', 1}, Map, Objects) -> lists:foldl(fun(Obj, Acc) -> [ Map(Obj) | Acc ] end, [], Objects);
```
Comment (fenollp) [inline applications/crossbar/src/cb_pager.erl]: Why not get arity with the `is_function/2` guard?

File: applications/crossbar/src/cb_pager.erl [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+apply_filter({'arity', 2}, Filter, Objects) -> lists:foldl(Filter, [], Objects).
+
+-spec last_key(list(), integer(), integer()) -> {integer()|undefined, list()}.
```
Comment (fenollp) [inline applications/crossbar/src/cb_pager.erl]: s / list() / kz_json:objects() /

File: applications/crossbar/src/cb_pager.erl [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    {_, LastKey, JObjs} =
+        lists:foldl(fun fold_query/2, {Limit, 'undefined', []}, [ {Db, View, CouchOpts, Filter} || Db <- MODbs ]),
+    {LastKey, JObjs}.
+
+-spec fold_query({binary(), binary(), list(), fun()}, {integer(), integer(), list()}) -> {integer(), integer(), list()}.
+fold_query({Db, View, CouchOpts, Filter}, {Limit, _LastKey, Res}) when is_integer(Limit), Limit > 0 ->
+    Queried = erlang:length(Res),
+    LimitWithLast = 1 + Limit - Queried,
+    DbResults = limited_query(LimitWithLast, Db, View, CouchOpts),
+    {LastKey, JObjs} = last_key(lists:reverse(DbResults), LimitWithLast, erlang:length(DbResults)),
```
Comment (fenollp) [inline applications/crossbar/src/cb_pager.erl]: Could we get better time complexity? `reverse` + `length` + more `reverse` in last_key's code

File: applications/crossbar/src/crossbar_filter.erl [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+build(QS, 'true') -> build_with(QS).
+
+-spec build_with(kz_json:object()) -> fun().
+build_with(QueryString) ->
+    fun(Doc) ->
+            filter_doc_by_querystring(Doc, QueryString)
+    end.
+
+-spec defined(cb_context:context()) -> boolean().
+defined(Context) ->
```
Comment (fenollp) [inline applications/crossbar/src/crossbar_filter.erl]: Could we rename this to `is_defined`?

File: applications/crossbar/src/cb_pager.erl (hunk 5 above)
Comment (jamhed) [inline applications/crossbar/src/cb_pager.erl]: actually, only one reverse there (other in last_key was a remnant).

File: applications/crossbar/src/crossbar_pager.erl [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                        {<<"page_size">>, PageSize},
+                        {<<"next_start_key">>, NextStartKey}
+                       ],
+                       JObj).
+
+-spec remove_paging(kz_json:object()) -> kz_json:object().
+remove_paging(JObj) ->
+    kz_json:delete_keys([<<"start_key">>, <<"page_size">>, <<"next_start_key">>], JObj).
+
+-spec format_response(cb_context:context(), integer(), integer()|'undefined', integer(), kz_json:objects()) -> cb_context:context().
```
Comment (fenollp) [inline applications/crossbar/src/crossbar_pager.erl]: `integer()|'undefined'` is `api_integer()`

File: applications/crossbar/src/cb_pager.erl [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+limited_query(Limit, Db, View, CouchOpts) ->
+    case kz_datamgr:get_results(Db, View, [{'limit', Limit} | CouchOpts]) of
+        {'ok', JObjs} -> JObjs;
+        {'error', 'not_found'} -> [];
+        {'error', Error} -> throw(Error)
+    end.
+
+-spec apply_filter(fun(), kz_json:objects()) -> kz_json:objects().
+apply_filter(Map, Objects) when is_function(Map, 1) ->
+    lists:foldl(fun(Obj, Acc) -> [ Map(Obj) | Acc ] end, [], Objects);
```
Comment (fenollp) [inline applications/crossbar/src/cb_pager.erl]: How about `[Map(Obj) || Obj <- Objects]`?

**Pull request closed without merging**